### DISCUSSION TITLE
BO-LIFTの利用

### DISCUSSION THREAD

**Discussion by n-yoshikawa** [souyakuchan/LLM_DD_Challenge#6] [Show and tell]
[BO-LIFT](https://github.com/ur-whitelab/BO-LIFT) というLLMを用いたベイズ最適化のライブラリにあるInverse designの機能を試してみました。
ライブラリの元となる論文は [Bayesian Optimization of Catalysts With In-context Learning](https://arxiv.org/abs/2304.05341) です。

```python
from rdkit import Chem
from sklearn.model_selection import train_test_split
import bolift

suppl = Chem.SDMolSupplier('CBLB_inhibitors_vsF.sdf')
X = []  # SMILES
y = []  # IC50 values

for mol in suppl:
    X.append(mol.GetProp('smiles'))
    IC50_range = mol.GetProp('IC50_range_nM')
    if '-' in IC50_range:
        IC50 = float(IC50_range.split('-')[1])
    else:
        IC50 = float(IC50_range[1:])
    y.append(IC50)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42)
# Create Gaussian process regression model
asktell = bolift.AskTellGPR()

# https://github.com/ur-whitelab/BO-LIFT/issues/18
for smiles, IC50 in zip(X_train[:-1], y_train[:-1]):
    asktell.tell(smiles, IC50, train=False)

asktell.tell(X_train[-1], y_train[-1])

# inverse design
asktell.inv_predict(10000)  #'\n\nCc1cc2C(N(Cc2c(C2CC2)n1)c1cccc(c1)C1(Cc2nncn2C)COC1)=O'
```

テキスト情報しか使っていないためそこまで精度は出ないように感じますが、こういうLLMの使い方もあるということで参考程度に見ていただければと思います。

**Comment by chronobcelp** [comment]
すいませんが、この通りにやりましたがエラーが出ました。
API Keyとか、何かコツとか必要ですか？

**Reply by n-yoshikawa** [reply]
こちらのコードも環境変数 OPENAI_API_KEY にAPI keyが設定されていることを仮定しています。

**Reply by chronobcelp** [reply]
環境変数 OPENAI_API_KEY にAPI keyは入力しました。
しかし、以下のようなメッセージが出ます。
「RateLimitError: You exceeded your current quota, please check your plan and billing details」

一応、ChatGPT Plusにしていますが、API使用はそれとは別の料金体系なんですね。調べてみます。

**Reply by souyakuchan** [reply]
料金体系が別でクレカ登録が必要で、API のほうは $5 分が無償付与されていますがそれを超えると止まります。なお GPT-4 API 利用は更に waitlist 登録して順番待ちが必要ですが、私のアカウントでは既に利用できるので、LLM創薬チャレンジ参加者には希望者に権限付与しています。

**Reply by n-yoshikawa** [reply]
ざっと調べたところ、APIの無料枠を使い切っている可能性が高そうです。 クレジットカードを登録して新しいAPI keyを発行すると解決するそうです https://stackoverflow.com/questions/75898276/

**Comment by chronobcelp** [comment]
soyakuchanにAPIをレンタルさせてもらいましたが、以下のエラーが出ます。
どうすればいいですか？

Cached embeddings not found. Creating new cache table.
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
[<ipython-input-43-6cff32027f99>](https://t9i2x0w8qqa-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20230531-060125-RC00_536671505#) in <cell line: 23>()
     22 # https://github.com/ur-whitelab/BO-LIFT/issues/18
     23 for smiles, IC50 in zip(X_train[:-1], y_train[:-1]):
---> 24     asktell.tell(smiles, IC50, train=False)
     25 
     26 asktell.tell(X_train[-1], y_train[-1])

3 frames
/usr/local/lib/python3.10/dist-packages/pydantic/main.cpython-310-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()

ValidationError: 1 validation error for OpenAI
__root__
  Parameters {'logit_bias'} should be specified explicitly. Instead they were passed in as part of `model_kwargs` parameter. (type=value_error)
